### PR TITLE
fix exact input payer bug

### LINKED ISSUE #87: payer in multihop should be router except for first swap
https://github.com/Uniswap/uniswap-v3-periphery/blob/8ad273f45cd42e04b5f8f15b6f4da57827122273/contracts/SwapRouter.sol#L145

i guess tests are passing because all tokens are approved

### PR DESCRIPTION
**Pull request by NoahZinsmeister**
closes #87

### CODE REVIEW

**Review by moodysalem — APPROVED** [review]

File: contracts/SwapRouter.sol [hunk 1]
```diff
@@ -142,10 +144,12 @@ contract SwapRouter is
                 0,
                 SwapCallbackData({
                     path: params.path.getFirstPool(), // only the first pool in the path is necessary
-                    payer: msg.sender
+                    payer: payer
                 })
             );
 
+            if (payer != address(this)) payer = address(this); // this contract pays for subsequent hops
```
Comment (moodysalem) [inline contracts/SwapRouter.sol:151]: wait this is an odd way to do it, it has a branch every time plus an mstore, we could just use a ternary in the payer?

Comment (moodysalem) [reply contracts/SwapRouter.sol:151]: if you just remove the if branch i'd guess it's cheaper and still correct

Comment (NoahZinsmeister) [reply contracts/SwapRouter.sol:151]: well you need something like this in any case. pseudocode:

hasPaid = false
// ...
payer: hasPaid ? address(this) : msg.sender
// ...
if (!hasPaid) hasPaid = false // if statement is so we only mstore once

so this is 2 MSTOREs, and 2 conditionals / loop iteration

my version is 2 MSTOREs, and 1 conditional / loop iteration

does that seem right?

Comment (moodysalem) [reply contracts/SwapRouter.sol:151]: i suspect the if statement creates a jump which is more expensive than just setting it every time

Comment (NoahZinsmeister) [reply contracts/SwapRouter.sol:151]: ok, will check the gas costs and merge whichever's cheaper

**Pull request merged**